# Analysis and visualization of MERFISH data with CellScopes.jl
[MERFISH](https://www.science.org/doi/10.1126/science.aaa6090) is an imaging-based spatial transcriptomics technique developped by Xiaowei Zhuang lab. MERFISH data are generated by a MERSCOPE platform from Vizgen.  Here we showcase how to use ```CellScopes.jl``` to analyze the MERFISH data. The example dataset in this tutorial is a human ovarian cancer dataset downloaded from the [Vizgen website](https://vizgen.com/data-release-program/?gclid=CjwKCAjw5MOlBhBTEiwAAJ8e1plC2ciW6OGOPpX44oZ5cnmk-7uM5L8ZD2BnW96PVzVRXbiAyCLLnhoC7JIQAvD_BwE).

## 1. Read the MERFISH data directly from MERSCOPE output
Like the case in other spatial modularity, we created a function ```read_merfish``` which conveniently reads the output files from MERSCOPE. Just provide the path to the MERSCOPE output and ```CellScope``` will take the rest of steps and generate a new spatial object type ```MerfishObject``` for downstream analysis.

```julia
import CellScopes as cs
merfish_dir = "/mnt/sdd/ovarian_cancer/"
merfish = cs.read_merfish(merfish_dir; prefix = "merfish")
```
This should create a ```MerfishObject``` object.
```
1.Loading transcript file...f
1.Transcript file loaded...
2.Reading cell polygons data...
Progress: 100%[==================================================] Time: 0:03:11
2.Cell polygons loaded!
3.Formatting cell polygons...
Progress: 100%[==================================================] Time: 0:00:02:51
3.Cell polygons formatted!
Adding prefix merfish to all cells...
MerfishObject in CellScopes.jl
Genes x Cells = 500 x 212425
Available data:
- rawCount
- normCount
- metaData
- spmetaData
- polynormCount
- coordData
- polygonData
All fields:
- rawCount
- normCount
- scaleCount
- metaData
- spmetaData
- varGene
- dimReduction
- clustData
- polyCount
- polynormCount
- coordData
- imputeData
- imageData
- polygonData
```

### 2. Normalization, dimension reduction and cell clustering
We then perform data normalization, dimension reduction and cell clustering on the Object using the general functions developed for all data types. Since MERSCOPE output does not include cell clutering and umap dimensional reduction, we generate these results using internal function in ```CellScopes```.
```julia
merfish = cs.normalize_object(merfish)
merfish = cs.scale_object(merfish)
merfish = cs.find_variable_genes(merfish; nFeatures=500)
merfish = cs.run_pca(merfish;  method=:svd, pratio = 0.9, maxoutdim = 20)
merfish = cs.run_umap(merfish; min_dist=0.4, n_neighbors=20)
merfish = cs.run_clustering(merfish; res=0.00005)
```

### 3. Visiualization
We provide a set of functions to help visualize the cell and gene expression on the dimension reduction space and the spatial space.

#### 3.1 Cell annotation

##### a. Visualize the cell annotation on umap.
Note that some outlier data points were usually seen in the MERFISH data. We set the axis limit to trim away these outlier points (usually only several points so not affect the overall results). We are working on including a similar clipping approach (e.g. clip.range parameter) from [Seurat](https://satijalab.org/seurat/articles/spatial_vignette_2.html) in the normalization step to solve the outlier datapoint issue in umap.
```julia
cs.dim_plot(merfish; dim_type = "umap", marker_size = 4, 
    x_lims = (-15, 14), y_lims = (-14, 14))
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/merfish_umap.png" width="600"> 

##### b. Visualize the cell annotation on tissue.
```julia
colors =["blue", "red","magenta", "seagreen", "deepskyblue3", 
    "green2", "purple","darkgoldenrod1", "cyan", "dodgerblue"]
celltypes = string.(unique(merfish.metaData.cluster))
anno_color=Dict(celltypes .=> colors)
cs.sp_dim_plot(merfish, "cluster"; anno_color = anno_color,
    do_label = false,marker_size = 5, 
    width=2500, height=2000, do_legend=false
    )
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/merfish_tissue.png" width="600"> 

##### c. Select a field of view to visualize the cell annotation
```julia
cs.plot_fov(merfish, 20, 20)
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/merfish_fov.png" width="600"> 

The grid plot above allows us to choose the specific region of interest for visualization. For instance, we can crop the view to focus on the hippocampus, as shown in the example below.

```julia
crop_fov = cs.subset_fov(merfish, [188, 189, 208, 209],20,20)
x1 = minimum(crop_fov.x)
x2 = maximum(crop_fov.x)
y1 = minimum(crop_fov.y)
y2 = maximum(crop_fov.y)
cs.sp_dim_plot(merfish, "cluster"; anno_color = anno_color,
    do_label = false,marker_size = 4, x_lims = (x1, x2),
    y_lims = (y1, y2),
    width=500, height=400, do_legend=false
    )
```

<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/merfish_crop2.png" width="600"> 

The function below is for visualizing the cell type annotation directly on cell polygons.
<br>
```julia
cs.plot_cell_polygons(merfish, "cluster"; cell_colors=colors, stroke_color="gray40",
    x_lims=(x1, x2), y_lims = (y1, y2),stroke_width=0.5,
    width = 500, height = 400)
```

<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/merfish_crop1.png" width="600"> 

#### 3.2 Gene expression visiualization
Plot genes on whole tissue:

```julia
cs.sp_feature_plot(merfish, ["COL1A1", "CDH1"]; 
    color_keys=["gray94", "lemonchiffon", "red"], 
    height=800, width=1000, marker_size = 4)
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/merfish_gene_whole.png" width="600"> 

Plot genes on the selected region:

```julia
cs.plot_gene_polygons(merfish, ["COL1A1", "CDH1"]; x_lims = (x1, x2),y_lims = (y1, y2),
    width = 500, height = 400, color_keys=["gray94", "lemonchiffon", "red","darkred"])
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/merfish_gene_polygon.png" width="600"> 

#### 3.3 Save the object

To save all data in the ```MerfishObject```, simply run:

```julia
cs.save(merfish; filename="merfish.jld2")
```
