# Analysis and visualization of Xenium data with CellScopes.jl
Here is the tutorial about how ```CellScopes.jl``` can be used to analyze the Xenium data. The dataset used for this tutorial is a mouse brain Xenium dataset available on the [10x website](https://www.10xgenomics.com/resources/datasets?query=&page=1&configure%5Bfacets%5D%5B0%5D=chemistryVersionAndThroughput&configure%5Bfacets%5D%5B1%5D=pipeline.version&configure%5BhitsPerPage%5D=500&configure%5BmaxValuesPerFacet%5D=1000&menu%5Bproducts.name%5D=In%20Situ%20Gene%20Expression).

## 1. Read the xenium data
Xenium data generated by the Xenium Analayzer can be easily imported into CellScopes by the function ```read_xenium```. The only parameter required by this function is the directory where the xenium data is stored in the local computer. When reading the data, the cells and genes can be filtered by specifying the parameters ```min_gene``` and ```min_cell```.

```julia
import CellScopes as cs
xenium_dir = "/mnt/sdc/new_analysis_cellscopes/xenium_mouse_brain/"
@time brain = cs.read_xenium(xenium_dir; min_gene = 0, min_cell = 0, prefix = "brain")
```
This should create a ```XeniumObject``` object.
```
XeniumObject in CellScopes.jl
Genes x Cells = 248 x 162033
Available data:
- rawCount
- metaData
- spmetaData
- coordData
All fields:
- rawCount
- normCount
- scaleCount
- metaData
- spmetaData
- varGene
- dimReduction
- clustData
- polyCount
- polynormCount
- coordData
- imputeData
- imageData
- polygonData
```

### 2. Normalization, dimension reduction and cell clustering
We then perform data normalization, dimension reduction and cell clustering on the Object.
```julia
brain = cs.normalize_object(brain)
brain = cs.scale_object(brain)
brain = cs.find_variable_genes(brain; nFeatures=200)
brain = cs.run_pca(brain;  method=:svd, pratio = 1, maxoutdim = 20)
```
### 3. (Optional) Use Baysor to generate cell polygons and assign gene expression to the polygons
The Xenium object already contained the cell polygon data generated by 10x Analyzer. If you want to use other tools (such as Baysor) to draw the cell polygons, please follow the step below. After the cell polygons were generated, we associate each polygon with the cell closest to it based on Euclidean distance by running ```polygons_cell_mapping```. Finally, you will need to assign gene expression values to the cell polygons with ```generate_polygon_counts```. **This step is entirely optional and can be omitted without impacting any of the subsequent analyses.**

```julia
import Baysor as B
scale=20
min_pixels_per_cell = 15
grid_step = scale / min_pixels_per_cell
bandwidth= scale / 10
count_molecules = deepcopy(brain.spmetaData.molecule)
count_molecules.cell[count_molecules.cell.==-1] .= 0
count_molecules.cell = string.(count_molecules.cell)
cell_baysor = [replace(x, "brain_" => "") for x in count_molecules.cell]
cell_baysor = parse.(Int64, cell_baysor)
polygons = B.boundary_polygons(count_molecules, cell_baysor, grid_step=grid_step, bandwidth=bandwidth)
brain.polygonData = polygons
brain = cs.polygons_cell_mapping(brain)
brain = cs.generate_polygon_counts(brain)
```
### 4. Visiualization
We provide a set of functions to help visualize the cell and gene expression on the dimension reduction space and the spatial space.

#### 4.1 Cell annotation
##### a. Visualize the cell annotation on tissue.
```julia
colors =["#4f8c9d", "#8dd2d8", "#0a4f4e", "yellow", "#229743", 
    "#8dd2d8", "#0a4f4e","#a7d64e", "#788c3b", "#57e24c", 
    "#683c00", "#f1bb99", "blue", "#9f1845", 
    "#f87197", "#ff0087", "#a57a6a", "#fabd2a", 
    "#374475", "#628df2", "#691b9e", "slateblue1", 
    "#d5d0fa", "black", "#ab7b05","#4f8c9d",
    "#8dd2d8", "#0a4f4e", "#4aeeb6", "#229743",
    "#a7d64e", "#788c3b", "#57e24c", "#683c00", 
    "#f1bb99", "#db3c18", "cyan", "#f87197", 
    "#ff0087", "#a57a6a", "#fabd2a", "#374475", 
    "#628df2", "#691b9e", "#b25aed", "red", 
    "fuchsia", "#ab7b05","orangered3", "#0a4f4e"]
celltypes = string.(unique(brain.metaData.cluster))
anno_color=Dict(celltypes .=> colors)
cs.sp_dim_plot(brain, "cluster"; anno_color = anno_color,
    do_label = false,marker_size = 5, 
    canvas_size = (2500, 2000), do_legend=false
    )
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/xenium_celltype.jpg" width="600"> 

##### b. Select a field of view to visualize the cell annotation
```julia
cs.plot_fov(brain, 20, 20)
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/xenium_fov.jpg" width="600"> 

The grid plot above allows us to choose the specific region of interest for visualization. For instance, we can crop the view to focus on the hippocampus, as shown in the example below.

```julia
hippo = cs.subset_fov(brain, [93, 96, 173, 176],20,20)
x1 = minimum(hippo.x)
x2 = maximum(hippo.x)
y1 = minimum(hippo.y)
y2 = maximum(hippo.y)
cs.sp_dim_plot(brain, "cluster"; anno_color = anno_color,
    do_label = false,marker_size = 7, x_lims = (x1, x2),
    y_lims = (y1, y2),
    canvas_size = (700, 600), do_legend=false
    )
```

<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/xenium_hippo1.png" width="600"> 

We can also visualize the cell type annotation directly on cell polygons.
<br>
```julia
cs.plot_cell_polygons(brain, "cluster"; x_lims = (x1, x2),
    y_lims = (y1, y2), cell_colors= colors, stroke_color="black",
    width = 800, height = 550)
```

<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/xeniium_hippo2.png" width="600"> 

#### 4.2 Gene expression visiualization
Plot genes on whole tissue:

```julia
cs.sp_feature_plot(brain, "Bcl11b"; color_keys=["gray94", "lemonchiffon", "red"], height=3000, width=3000, marker_size = 4)
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/xenium_whole_gene.jpg" width="600"> 

Plot genes on the selected region:

```julia
cs.plot_gene_polygons(brain, ["Bcl11b"]; x_lims = (x1, x2),y_lims = (y1, y2),
    width = 800, height = 550, color_keys=["#440154", "#440154","#3b528b","#ffff67"], bg_color="black")
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/xenium_gene.png" width="600"> 

#### 4.3 Gene imputation
```CellScopes.jl``` provides wrapper functions to run gene imputation using three different tools: SpaGE, gimVI, and tangram. The code snippet below demonstrates how to run SpaGE using ```CellScopes.jl```.

```julia
data_path = "/mnt/sdc/new_analysis_cellscopes/brain_sc/"
SpaGE_path = "/mnt/sdc/new_analysis_cellscopes/SpaGE/"
brain = cs.run_spaGE(brain, data_path, SpaGE_path)
```
Imputed gene values can be visualized by setting ```use_imputed = true``` in the ```sp_feature_plot``` function.

```julia
cs.sp_feature_plot(brain, ["Adcy5","Ccdc3", "Prox1"]; color_keys=["gray90", "lemonchiffon", "red"], use_imputed = true)
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/xenium_gene_impute1.png" width="600"> 

To zoom in the area of interest, simply run: 

```julia
cs.sp_feature_plot(brain2, ["Ccdc3"]; 
    color_keys=["gray80", "lemonchiffon", "red3"], order=true,
    use_imputed = true, x_lims = (x1, x2), marker_size = 10,
    y_lims = (y1, y2),width = 800, height = 550)
```
<img src="https://github.com/HaojiaWu/CellScopes.jl/blob/main/data/xenium_gene_impute2.png" width="600"> 

#### 4.4 Save the object

After completion of the analysis, you can save the xenium data object.

```julia
cs.save(brain; filename = "brain_final.jld2")
```





